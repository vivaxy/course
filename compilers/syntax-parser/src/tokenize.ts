/**
 * @since 2020-02-26 02:39
 * @author vivaxy
 */
interface Pattern {
  type: string;
  regExps: RegExp[];
  ignore?: boolean;
}

function getMatched(patterns: Pattern[], code: string) {
  for (const pattern of patterns) {
    for (const regExp of pattern.regExps) {
      const formattedRegExp = new RegExp(`^(${regExp.source})`);
      const matched = code.match(formattedRegExp);
      if (matched) {
        return {
          type: pattern.type,
          regExp: regExp,
          ignore: pattern.ignore,
          value: matched[1],
        };
      }
    }
  }
  throw new Error('Unexpected token');
}

export default function tokenize(patterns: Pattern[], code: string) {
  let index = 0;
  const tokens = [];
  while (index < code.length) {
    const matched = getMatched(patterns, code.slice(index));
    if (matched.ignore) {
      index += matched.value.length;
    } else {
      tokens.push({
        type: matched.type,
        value: matched.value,
        position: [index, (index += matched.value.length)],
      });
    }
  }
  return tokens;
}
