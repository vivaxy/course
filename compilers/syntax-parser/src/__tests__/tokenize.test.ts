/**
 * @since 2020-02-26 03:57
 * @author vivaxy
 */
import tokenize from '../tokenize';

test('', function () {
  const stateMachine = {};
  const code = ``;
  const tokens = tokenize(stateMachine, code);
  const expected = [];
  expect(tokens).toBe(expected);
})
