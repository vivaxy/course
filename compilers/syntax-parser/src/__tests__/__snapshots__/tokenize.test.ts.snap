// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`binary operation 1`] = `
Array [
  Object {
    "position": Array [
      0,
      1,
    ],
    "type": "Number",
    "value": "1",
  },
  Object {
    "position": Array [
      2,
      3,
    ],
    "type": "Operator",
    "value": "+",
  },
  Object {
    "position": Array [
      4,
      5,
    ],
    "type": "Number",
    "value": "2",
  },
]
`;

exports[`binary operation 2`] = `
Array [
  Object {
    "position": Array [
      0,
      1,
    ],
    "type": "Number",
    "value": "1",
  },
  Object {
    "position": Array [
      2,
      3,
    ],
    "type": "Operator",
    "value": "+",
  },
  Object {
    "position": Array [
      4,
      5,
    ],
    "type": "Number",
    "value": "2",
  },
  Object {
    "position": Array [
      6,
      7,
    ],
    "type": "Operator",
    "value": "-",
  },
  Object {
    "position": Array [
      8,
      9,
    ],
    "type": "Number",
    "value": "3",
  },
]
`;

exports[`variable declaration 1`] = `
Array [
  Object {
    "position": Array [
      0,
      3,
    ],
    "type": "Declarator",
    "value": "var",
  },
  Object {
    "position": Array [
      4,
      5,
    ],
    "type": "Identifier",
    "value": "a",
  },
  Object {
    "position": Array [
      6,
      7,
    ],
    "type": "Operator",
    "value": "=",
  },
  Object {
    "position": Array [
      8,
      9,
    ],
    "type": "Identifier",
    "value": "b",
  },
]
`;
